src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,545 @@
+            if peft_config.eva_config.use_label_mask and hasattr(inputs, "labels"):
+                mask = torch.logical_and(mask, inputs["labels"] != peft_config.eva_config.label_mask_value)
+            return mask.nonzero()
+            ```
+        prepare_layer_inputs_fn (Union[callable, Dict[str, callable], None]):
+            If a layer receives multiple inputs as a list, per default the first input is used. This function can be
+            used to modify this behaviour. Accepts a dictionary with layer names as keys. Default logic:
+            ```
+            if isinstance(inputs, torch.Tensor):
+                return inputs
+            elif isinstance(inputs, (tuple, list)):
+                return inputs[0]
+            else:
+                raise ValueError(...)
+            ```
+        adapter_name (str): The name of the adapter to initialize the weights for.
+        show_progress_bar (bool): Whether to show a progress bar. Default is True.
+
+    Returns:
+        model (torch.nn.Module): The model with the initialized LoRA weights.
+    """
+    if not hasattr(model, "peft_config"):
+        raise ValueError("model must be a PeftModel")
+    peft_config = model.peft_config[adapter_name]
+
+    # eva currently only works with a single active adapter
+    if len(model.active_adapters) > 1:
+        raise ValueError("`initialize_lora_eva_weights` currently only works with a single active adapter")
+
+    # initialize_lora_eva_weights only works with `init_lora_weights='eva'`
+    if peft_config.init_lora_weights != "eva":
+        raise ValueError("`initialize_lora_eva_weights` can only be used with `init_lora_weights='eva'`")
+
+    # compute svd
+    with model.disable_adapter():
+        eva_state_dict = get_eva_state_dict(
+            model=model,
+            peft_config=peft_config,
+            dataloader=dataloader,
+            forward_fn=forward_fn,
+            get_indices_fn=get_indices_fn,
+            prepare_layer_inputs_fn=prepare_layer_inputs_fn,
+            show_progress_bar=show_progress_bar,
+        )
+
+    # assert all lora layers are contained in eva_state_dict
+    update_layer_kwargs = {
+        "adapter_name": adapter_name,
+        "lora_alpha": peft_config.lora_alpha,
+        "lora_dropout": peft_config.lora_dropout,
+        "use_rslora": peft_config.use_rslora,
+        "use_dora": peft_config.use_dora,
+    }
+    missing_eva_inits = []
+    for name, module in model.named_modules():
+        if not isinstance(module, LoraLayer):
+            continue
+        if name in eva_state_dict:
+            w = eva_state_dict.pop(name)
+            module.update_layer(r=w.size(0), init_lora_weights=peft_config.init_lora_weights, **update_layer_kwargs)
+            module.lora_A[adapter_name].weight.copy_(w)
+        else:
+            module.update_layer(r=peft_config.r, init_lora_weights=True, **update_layer_kwargs)
+            missing_eva_inits.append(name)
+            print(name, type(module))
```

Comment:
Add a test for this `test_missing_eva_inits`. Since EVA at the moment does not support embedding layers for example, we are doing standard lora init if such layers have adapters.